### PR TITLE
Fix CRF training

### PR DESCRIPTION
**Pull request by ndrewl**
This fixes gradient backpropagation and weights update in the CRF layer.

There was excessive addition to gradients computed at the first step which caused overshooting, and therefore training instabilities.
The addition was produced by a misplaced softmax operation which doesn't preserve zeros.
In this PR we skip backpropagation and weights updates related to transitions for the first (possibly the only) step where we don't have such information.

### CODE REVIEW

File: NeoML/src/Dnn/Layers/CrfLayer.cpp [hunk 1]
```diff
@@ -123,49 +123,44 @@ bool CCrfCalculationLayer::isFirstStep() const
...
 	CConstFloatHandle currentProbabilities = inputBlobs[I_ClassLogProb]->GetData();
-	// The vector of partial function value on the previous step (alpha in the forward-backward algorithm, forward pass)
-	// Contains zeros for the first step
-	CConstFloatHandle sequenceProbabilities = inputBlobs[I_ClassSeqLogProb]->GetData();
-	tempSumBlob->Clear();
-	CFloatHandle tempSum = tempSumBlob->GetData();
-	if( !isFirstStep() ) {
+
+	if( isFirstStep() ) {
+		// On the first step all we have to do is initialize the O_ClassSeqLogProb output with current element probabilities
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/CrfLayer.cpp]: Why do we skip `outputBlobs[O_BestPrevClass]` here? (if `!IsLearningPerformed()`)

M.b. we should fill it based on `outputBlobs[O_BestPrevClass]` here?

File: NeoML/src/Dnn/Layers/CrfLayer.cpp [hunk 2]
```diff
@@ -123,49 +123,44 @@ bool CCrfCalculationLayer::isFirstStep() const
...
 			numberOfClasses * numberOfClasses, Transitions()->GetData());
 		// Add the partial function values calculated above
-		// Adding row by row except for the first step when they are equal to 0
 		MathEngine().AddVectorToMatrixRows(batchWidth, tempSum, tempSum,
 			numberOfClasses, numberOfClasses, sequenceProbabilities);
-	}
 
-	if(!IsLearningPerformed()) {
-		// Add the estimates (logits) of the current element classes
-		MathEngine().AddVectorToMatrixColumns(tempSum, tempSum,
```
Comment (FedyuninV) [inline NeoML/src/Dnn/Layers/CrfLayer.cpp:150]: Why this call was removed?

Comment (ndrewl) [reply NeoML/src/Dnn/Layers/CrfLayer.cpp:150]: It's not removed, it's now simplified and moved to line 162. As you see, the following argmax is done over rows. So adding a vector `V` to columns of a matrix `M` doesn't affect the result: `argmax(M + V) = argmax(M)`. `O_ClassSeqLogProb` remains correct as well (the max is done over rows): `max(M + V) = V + max(M)`. It's a little optimization, instead of doing broadcast over columns here, we add vector `currentProbabilites` in the end, and get the same result.

File: NeoML/src/Dnn/Layers/CrfLayer.cpp (hunk 1 above)
Comment (ndrewl) [reply NeoML/src/Dnn/Layers/CrfLayer.cpp]: Do you mean use of `outputBlobs[O_BestPrevClass] != 0` instead of `!IsLearningPerformed()` to compute `O_BestPrevClass` only when it is connected?
That's a good point. Actually, I have another PR about computing `O_BestPrevClass` and making `CBestSequenceLayer` usable during training, let's discuss it there. I'll publish it right after this one gets merged.

**Review by FedyuninV — APPROVED** [review]
